From: Jon Severinsson <jon@severinsson.net>
Date: Sun, 20 Jul 2014 20:00:00 +0200
Subject: Add support for rcS.d init scripts to the sysv-generator

As we unconditionally add After=local-fs.target now, this requires some
adjustments to the test suite.
---
 man/systemd.special.xml                   |  5 +++++
 src/insserv-generator/insserv-generator.c |  2 +-
 src/sysv-generator/sysv-generator.c       | 31 +++++++++++++++++++++++++++----
 test/sysv-generator-test.py               | 10 +++++-----
 4 files changed, 38 insertions(+), 10 deletions(-)

diff --git a/man/systemd.special.xml b/man/systemd.special.xml
index 26974ed..611313b 100644
--- a/man/systemd.special.xml
+++ b/man/systemd.special.xml
@@ -345,6 +345,11 @@
           target unit for those mounts listed in
           <filename>/etc/fstab</filename> that have the
           <option>auto</option> mount option set.</para>
+
+          <para>systemd automatically adds dependencies of type
+          <varname>After=</varname> for this target unit to all SysV
+          init script service units with an LSB header referring to the
+          <literal>$local_fs</literal> facility.</para>
         </listitem>
       </varlistentry>
       <varlistentry>
diff --git a/src/insserv-generator/insserv-generator.c b/src/insserv-generator/insserv-generator.c
index 58c1ed2..7f1ada0 100644
--- a/src/insserv-generator/insserv-generator.c
+++ b/src/insserv-generator/insserv-generator.c
@@ -65,7 +65,7 @@ static int sysv_translate_facility(const char *name, const char *filename, char
 
         static const char * const table[] = {
                 /* LSB defined facilities */
-                "local_fs",             NULL,
+                "local_fs",             SPECIAL_LOCAL_FS_TARGET,
                 "network",              SPECIAL_NETWORK_TARGET,
                 "named",                SPECIAL_NSS_LOOKUP_TARGET,
                 "portmap",              SPECIAL_RPCBIND_TARGET,
diff --git a/src/sysv-generator/sysv-generator.c b/src/sysv-generator/sysv-generator.c
index fe4bbee..fe1cbdd 100644
--- a/src/sysv-generator/sysv-generator.c
+++ b/src/sysv-generator/sysv-generator.c
@@ -43,6 +43,7 @@
 #include "util.h"
 
 typedef enum RunlevelType {
+        RUNLEVEL_SYSINIT,
         RUNLEVEL_UP,
         RUNLEVEL_DOWN
 } RunlevelType;
@@ -52,6 +53,9 @@ static const struct {
         const char *target;
         const RunlevelType type;
 } rcnd_table[] = {
+        /* Debian SysV runlevel for early boot */
+        { "rcS.d",  SPECIAL_SYSINIT_TARGET,   RUNLEVEL_SYSINIT },
+
         /* Standard SysV runlevels for start-up */
         { "rc1.d",  SPECIAL_RESCUE_TARGET,     RUNLEVEL_UP },
         { "rc2.d",  SPECIAL_MULTI_USER_TARGET, RUNLEVEL_UP },
@@ -76,6 +80,7 @@ typedef struct SysvStub {
         char *name;
         char *path;
         char *description;
+        bool sysinit;
         int sysv_start_priority;
         char *pid_file;
         char **before;
@@ -192,6 +197,8 @@ static int generate_unit_file(SysvStub *s) {
 
         if (s->description)
                 fprintf(f, "Description=%s\n", s->description);
+        if (s->sysinit)
+                fprintf(f, "DefaultDependencies=no\n");
 
         STRV_FOREACH(p, s->before)
                 fprintf(f, "Before=%s\n", *p);
@@ -206,11 +213,12 @@ static int generate_unit_file(SysvStub *s) {
                 "\n[Service]\n"
                 "Type=forking\n"
                 "Restart=no\n"
-                "TimeoutSec=5min\n"
+                "TimeoutSec=%s\n"
                 "IgnoreSIGPIPE=no\n"
                 "KillMode=process\n"
                 "GuessMainPID=no\n"
                 "RemainAfterExit=%s\n",
+                s->sysinit ? "0" : "5min",
                 yes_no(!s->pid_file));
 
         if (s->pid_file)
@@ -274,13 +282,19 @@ static int sysv_translate_facility(const char *name, const char *filename, char
 
         static const char * const table[] = {
                 /* LSB defined facilities */
-                "local_fs",             NULL,
+                "local_fs",             SPECIAL_LOCAL_FS_TARGET,
                 "network",              SPECIAL_NETWORK_ONLINE_TARGET,
                 "named",                SPECIAL_NSS_LOOKUP_TARGET,
                 "portmap",              SPECIAL_RPCBIND_TARGET,
                 "remote_fs",            SPECIAL_REMOTE_FS_TARGET,
-                "syslog",               NULL,
+                "syslog",               "systemd-journald-dev-log.socket",
                 "time",                 SPECIAL_TIME_SYNC_TARGET,
+                /* special rcS init scripts */
+                "cryptdisks",           "cryptsetup.target",
+                "mountall",             SPECIAL_LOCAL_FS_TARGET,
+                "mountnfs",             SPECIAL_REMOTE_FS_TARGET,
+                "checkroot",            "systemd-remount-fs.service",
+                "dbus",                 "dbus.socket",
         };
 
         char *filename_no_sh, *e, *m;
@@ -711,6 +725,11 @@ static int fix_order(SysvStub *s, Hashmap *all_services) {
                 if (s->has_lsb && other->has_lsb)
                         continue;
 
+                /* Don't order units between sysinit and regular sysv services,
+                 * they are ordered before and after basic.target anyway. */
+                if (other->sysinit != s->sysinit)
+                        continue;
+
                 if (other->sysv_start_priority < s->sysv_start_priority) {
                         r = strv_extend(&s->after, other->name);
                         if (r < 0)
@@ -911,8 +930,12 @@ static int set_dependencies_from_rcnd(const LookupPaths *lp, Hashmap *all_servic
                                 }
 
                                 if (de->d_name[0] == 'S')  {
+                                        if (rcnd_table[i].type == RUNLEVEL_SYSINIT) {
+                                                service->sysinit = true;
+                                                service->sysv_start_priority = a*10 + b;
+                                        }
 
-                                        if (rcnd_table[i].type == RUNLEVEL_UP)
+                                        if (rcnd_table[i].type == RUNLEVEL_UP && !service->sysinit)
                                                 service->sysv_start_priority = MAX(a*10 + b, service->sysv_start_priority);
 
                                         r = set_ensure_allocated(&runlevel_services[i], NULL);
diff --git a/test/sysv-generator-test.py b/test/sysv-generator-test.py
index 838dd57..f92bf25 100755
--- a/test/sysv-generator-test.py
+++ b/test/sysv-generator-test.py
@@ -186,7 +186,7 @@ class SysvGeneratorTest(unittest.TestCase):
         # $local_fs does not need translation, don't expect any dependency
         # fields here
         self.assertEqual(set(s.options('Unit')),
-                         set(['Documentation', 'SourcePath', 'Description']))
+                         set(['Documentation', 'SourcePath', 'Description', 'After']))
 
         self.assertEqual(s.get('Service', 'Type'), 'forking')
         init_script = os.path.join(self.init_d_dir, 'foo')
@@ -262,9 +262,9 @@ class SysvGeneratorTest(unittest.TestCase):
                          ['may1.service', 'must1.service', 'must2.service', 'ne_may2.service'])
 
         # other services should not depend on each other
-        self.assertFalse(results['must1.service'].has_option('Unit', 'After'))
-        self.assertFalse(results['must2.service'].has_option('Unit', 'After'))
-        self.assertFalse(results['may1.service'].has_option('Unit', 'After'))
+        self.assertEqual(results['must1.service'].get('Unit', 'After'), 'local-fs.target')
+        self.assertEqual(results['must2.service'].get('Unit', 'After'), 'local-fs.target')
+        self.assertEqual(results['may1.service'].get('Unit', 'After'), 'local-fs.target')
 
     def test_symlink_prio_deps(self):
         '''script without LSB headers use rcN.d priority'''
@@ -294,7 +294,7 @@ class SysvGeneratorTest(unittest.TestCase):
         err, results = self.run_generator()
         self.assertEqual(list(results), ['foo.service'])
         self.assertEqual(set(results['foo.service'].options('Unit')),
-                         set(['Documentation', 'SourcePath', 'Description']))
+                         set(['Documentation', 'SourcePath', 'Description', 'After']))
         # should create symlinks for the alternative names
         for f in ['bar.service', 'baz.service']:
             self.assertEqual(os.readlink(os.path.join(self.out_dir, f)),
